Add support for CCS811 digital gas sensor

Mani-Sadhasivam opened the pull request:
This PR adds support for AMS CCS811 digital gas sensor. Sensor is capable of reporting the following parameters:

1. Co2 concentration
2. VOC concentration
3. Sensor Voltage
4. Sensor Current

Relevant channels have been added to the sensor API and a sample application is provided for demonstrating the driver usage.

The driver works in adjacent with #5562 and #5564

Mani-Sadhasivam committed:
sensor: Add Co2, VOC, Voltage, Current channels

CCS811 digital gas sensor supports the following channels:

1. Co2 Channel
2. VOC Channel
3. Voltage Channel
4. Current Channel

Add support for the those in API.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

Mani-Sadhasivam committed:
drivers: sensor: Add support for CCS811 Digital Gas Sensor

Add support for AMS CCS811 Digital Gas Sensor for monitoring
indoor air quality. This sensor reports the following parameters:

1. Co2 concentration
2. VOC concentration
3. Sensor voltage
4. Sensor current

This driver only supports polling mode as of now.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

Mani-Sadhasivam committed:
samples: sensor: Add sample application for CCS811 sensor

Add sample application for AMS CCS811 digital gas sensor driver.
This application fetches Co2, VOC, Voltage and Current readings
from the sensor and prints it on serial console continuously.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

jfischer-phytec-iot commented on drivers/sensor/ccs811/Kconfig:
```diff
@@ -0,0 +1,68 @@
+	hex
+	prompt "I2C address for CCS811 Sensor"
+	depends on CCS811
+	default "0x5B"
```
Please add `range 0x5a 0x5b`

Mani-Sadhasivam replied:
Ack

jfischer-phytec-iot commented on drivers/sensor/ccs811/Kconfig:
```diff
@@ -0,0 +1,68 @@
+	help
+	  I2C address of the CCS811 sensor.
+
+	  0x5A: ADDR pin connected to GND.
```
Remove empty line L32. Please fix help about address selection, e.g. 0x5a: I2C_ADDR connected to GND, 0x5b: I2C_ADDR connected to VDD

Mani-Sadhasivam replied:
Ack

jfischer-phytec-iot commented on drivers/sensor/ccs811/ccs811.c:
```diff
@@ -0,0 +1,218 @@
+		SYS_LOG_ERR("Failed to set Measurement mode");
+		return -EIO;
+	}
+
```
It would be good to check the CCS811_STATUS  for errors after the drive mode is selected.

Mani-Sadhasivam replied:
Since drive mode is not user configurable, I didn't thought about implementing the error check. It won't hurt to add it though ;-)

ydamigos commented:
Could you also add a device tree binding for the sensor?

Mani-Sadhasivam commented:
@jfischer-phytec-iot @ydamigos Incorporated your suggestions as below:

1. Modified Kconfig to add I2C address range
2. Added error check after setting measurement mode
3. Added YAML bindings for the sensor

ydamigos reviewed (commented):
I am sorry, I wasn't clear enough. My suggestion was to add a device tree binding for the sensor, in order to use it in your sample using a overlay file (like it is done in sample bluetooth/hci_uart) and get configuration like CCS811_NAME, CCS811_I2C_ADDR and CCS811_I2C_MASTER_DEV_NAME from DT. If you are not planning to add such functionality in your sample, it is better to remove the DT binding.

ydamigos commented on drivers/sensor/ccs811/Kconfig:
```diff
@@ -0,0 +1,69 @@
+	string
+	prompt "GPIO device"
+	default "GPIOC"
+	depends on CCS811
```
Shouldn't this depend on CCS811_GPIO_WAKEUP?

Mani-Sadhasivam replied:
I have plans to add interrupt support also. So, just made it independent of GPIO_WAKEUP so that it can be used for both.

Mani-Sadhasivam replied:
Anyway will make it dependent on GPIO_WAKEUP for now!

ydamigos replied:
> I have plans to add interrupt support also. So, just made it independent of GPIO_WAKEUP so that it can be used for both.

I believe, it is more flexible to use different GPIO_DEV_NAME for wakeup and interrupt.

Mani-Sadhasivam commented:
@ydamigos Removed the binding and made GPIO_DEV_NAME dependent on GPIO_WAKEUP.

ydamigos reviewed (commented):
I am fine with the changes. Unfortunately, I don't have the hardware to test sensor driver.

Mani-Sadhasivam commented:
@ydamigos Thanks for the review. I have tested this driver with CCS811 connected to 96Boards Carbon.

ydamigos commented on drivers/sensor/ccs811/ccs811.c:
```diff
@@ -0,0 +1,230 @@
+
+	/* Switch device to application mode */
+	ret = switch_to_app_mode(drv_data->i2c);
+	if (ret)
```
I missed this one. Please add brackets.

Mani-Sadhasivam replied:
Ah... Fixed this one as well as one more. Thanks for pointing out!

nashif commented:
please resolve conflict

Mani-Sadhasivam commented:
@nashif Done!
